File: 432_alloone_data_structure/main.js
var AllOne = /** @class */ (function () {
    function AllOne() {
        this.counter = new Map();
        this.words = new Map();
    }
    AllOne.prototype.inc = function (key) {
        if (this.counter.get(key) == undefined) {
            this.counter.set(key, 0);
        }
        var oldCounter = this.counter.get(key);
        if (this.words.get(oldCounter) == undefined) {
            this.words.set(oldCounter, new Map());
        }
        this.words.get(oldCounter).delete(key);
        this.counter.set(key, oldCounter + 1);
        if (this.words.get(this.counter.get(key)) == undefined) {
            this.words.set(this.counter.get(key), new Map());
        }
        this.words.get(this.counter.get(key)).set(key, true);
    };
    AllOne.prototype.dec = function (key) {
        var oldCounter = this.counter.get(key);
        this.words.get(oldCounter).delete(key);
        this.counter.set(key, oldCounter - 1);
        this.words.get(this.counter.get(key)).set(key, true);
    };
    AllOne.prototype.getMaxKey = function () {
        var n = this.words.size;
        var res = "";
        if (this.words.get(n - 1).size > 0) {
            res = this.words.get(n - 1).keys[0];
        }
        return res;
    };
    AllOne.prototype.getMinKey = function () {
        var n = this.words.size;
        var res = "";
        for (var i = 0; i < n; i++) {
            if (this.words.get(i).size > 0) {
                res = this.words.get(i).keys[0];
                break;
            }
        }
        return res;
    };
    return AllOne;
}());
var obj = new AllOne();
obj.inc("hello");
obj.inc("hello");
obj.getMaxKey();
obj.getMinKey();
obj.inc("leet");
obj.getMaxKey();
obj.getMinKey();

File: 432_alloone_data_structure/main.go
package main

import "fmt"

func main() {
	obj := Constructor()
	obj.Inc("hello")
	obj.Inc("hello")
	fmt.Println(obj.GetMaxKey())
	fmt.Println(obj.GetMinKey())
	obj.Inc("leet")
	obj.Dec("hello")
	fmt.Println(obj.GetMaxKey())
	fmt.Println(obj.GetMinKey())
}

type AllOne struct {
	counter map[string]int
	words   map[int]map[string]struct{}
}

func Constructor() AllOne {
	return AllOne{
		counter: map[string]int{},
		words:   map[int]map[string]struct{}{},
	}
}

func (this *AllOne) Inc(key string) {
	oldCounter, ok := this.counter[key]
	if !ok {
		this.counter[key] = 0
		oldCounter = 0
	}

	delete(this.words[oldCounter], key)
	this.counter[key] += 1
	if this.words[this.counter[key]] == nil {
		this.words[this.counter[key]] = map[string]struct{}{}
	}
	this.words[this.counter[key]][key] = struct{}{}
}

func (this *AllOne) Dec(key string) {
	oldCounter := this.counter[key]

	delete(this.words[oldCounter], key)
	this.counter[key] -= 1

	if this.counter[key] == 0 {
		return
	}
	if this.words[this.counter[key]] == nil {
		this.words[this.counter[key]] = map[string]struct{}{}
	}
	this.words[this.counter[key]][key] = struct{}{}
}

func (this *AllOne) GetMaxKey() string {
	if len(this.words) == 0 {
		return ""
	}

	for i := len(this.words); i > 0; i-- {
		for j := range this.words[i] {
			return j
		}
	}

	return ""
}

func (this *AllOne) GetMinKey() string {
	if len(this.words) == 0 {
		return ""
	}

	for i := 1; i <= len(this.words); i++ {
		for j := range this.words[i] {
			return j
		}
	}

	return ""
}
